src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers.Binary;
+using System.Globalization;
+
+namespace System.Security.Cryptography.Encoding.Tests.Cbor
+{
+    internal partial class CborReader
+    {
+        public CborTag ReadTag()
+        {
+            CborTag tag = PeekTagCore(out int additionalBytes);
+
+            AdvanceBuffer(1 + additionalBytes);
+            _isTagContext = true;
+            return tag;
+        }
+
+        public void ReadTag(CborTag expectedTag)
+        {
+            CborTag tag = PeekTagCore(out int additionalBytes);
+
+            if (expectedTag != tag)
+            {
+                throw new InvalidOperationException("CBOR tag mismatch.");
+            }
+
+            AdvanceBuffer(1 + additionalBytes);
+            _isTagContext = true;
+        }
+
+        public CborTag PeekTag() => PeekTagCore(out int _);
+
+        private CborTag PeekTagCore(out int additionalBytes)
+        {
+            CborInitialByte header = PeekInitialByte(expectedType: CborMajorType.Tag);
+            return (CborTag)ReadUnsignedInteger(_buffer.Span, header, out additionalBytes);
+        }
+
+        // Additional tagged type support
+
+        public DateTimeOffset ReadDateTimeOffset()
+        {
+            // implements https://tools.ietf.org/html/rfc7049#section-2.4.1
+
+            switch (ReadTag())
```

Comment:
Note that unlike the primitive reader methods, reader implementations for tagged values might partially advance the reader state if there is a format error (note that the same is also true for the `SkipValue()` method).

I'm wondering whether this might justify using some sort of checkpointing logic so that the reader state can be rolled back in case of an error, but I'm not sure whether the benefits would justify the added complexity here.

Reply:
The only time SkipValue can fail is when the content is corrupt and moving forward again would not succeed.

It should generally be true that calling the wrong read method does not corrupt the reader (or, if you prefer, any wrong read call utterly destroys the reader).

So either this should avoid advancing at all until the end (which probably requires passing some offsets into some helpers that previously didn't need them), or the body should be something like

```C#
ReadOnlyMemory<byte> checkpoint = _buffer;

try
{
    // current code here
}
catch
{
    _buffer = checkpoint;
    throw;
}
```

Reply:
It probably would need a bit more work than that in the general case (e.g. resetting counters, popping items off the stack, etc). Wondering whether it might be worth exposing as protected methods for folks extending the Reader with new types, i.e.
```csharp
protected SomeOpaqueCheckpointToken CreateCheckpoint();
protected RestoreCheckpoint(SomeOpaqueCheckpointToken token);
```

Reply:
Assuming we're ok with state changes on format errors, it might be possible to dodge the above monstrosity and just use `PeekTag()`, at least for DateTime.